src/tools/piv-tool.c

```diff
@@ -337,27 +338,33 @@ static int gen_key(const char * key_info)
 
 #if OPENSSL_VERSION_NUMBER < 0x30000000L
 	evpkey = EVP_PKEY_new();
+	if (!evpkey) {
+		sc_log_openssl(ctx);
+		fprintf(stderr, "allocation of key failed\n");
+		return r;
+	}
 #endif
 
 	if (keydata.key_bits > 0) { /* RSA key */
 		BIGNUM *newkey_n, *newkey_e;
 #if OPENSSL_VERSION_NUMBER < 0x30000000L
-		RSA * newkey = NULL;
-		newkey = RSA_new();
-		if (newkey == NULL) {
+		RSA *newkey = RSA_new();
+		if (!newkey) {
+			sc_log_openssl(ctx);
 			EVP_PKEY_free(evpkey);
 			free(keydata.pubkey);
 			free(keydata.exponent);
 			fprintf(stderr, "gen_key RSA_new failed %d\n",r);
 			return -1;
 		}
 #else
-		EVP_PKEY_CTX *ctx = NULL;
+		EVP_PKEY_CTX *cctx = NULL;
 		OSSL_PARAM_BLD *bld = NULL;
 		OSSL_PARAM *params = NULL;
 #endif
 
 		if (!keydata.pubkey || !keydata.exponent) {
+			sc_log_openssl(ctx);
```

Comment:
I think in this case no openssl operation is involved. These data come from the card.